kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java

```diff
@@ -98,23 +103,66 @@ public Mono<Set<String>> extract(AccessControlService acs, Object value, Map<Str
         .bodyToMono(new ParameterizedTypeReference<>() {});
     //@formatter:on
 
-    return userOrganizations
-        .map(orgsMap -> {
-          var groupsByOrg = acs.getRoles()
-              .stream()
-              .filter(role -> role.getSubjects()
+    Map<Role, List<Subject>> teamsMap = new HashMap<>();
+    acs.getRoles().forEach(r -> {
+      List<Subject> subjects = r.getSubjects().stream()
+          .filter(s -> s.getProvider().equals(Provider.OAUTH_GITHUB) && s.getType().equals("team")).toList();
+      teamsMap.put(r, subjects);
+    });
+
+    Flux<Pair<Role, Object>> teams = Flux.fromIterable(acs.getRoles())
+        .flatMap(role -> Flux.fromIterable(role.getSubjects())
+            .filter(s -> s.getProvider().equals(Provider.OAUTH_GITHUB) && s.getType().equals("team"))
+            .flatMap(subject -> {
+              String orgTeam = subject.getValue();
+              if (orgTeam.contains("/")) {
+                String[] organizationAndTeam = orgTeam.split("/");
+                String org = organizationAndTeam[0];
+                String team = organizationAndTeam[1];
+                WebClient wc = WebClient.create(
+                    String.format(GITHUB_TEAMS_URL, org, team, username));
+
+                return wc.get()
+                    .headers(headers -> {
+                      headers.set(HttpHeaders.ACCEPT, GITHUB_ACCEPT_HEADER);
+                      OAuth2UserRequest request = (OAuth2UserRequest) additionalParams.get("request");
+                      headers.setBearerAuth(request.getAccessToken().getTokenValue());
+                    })
+                    .retrieve()
+                    .bodyToMono(new ParameterizedTypeReference<>() {})
+                    .onErrorResume(Exception.class, ex -> Mono.just(new HashMap<String, String>()))
+                    .map(m -> Pair.of(role, m))
+                    .flux();
+              }
+              return Flux.empty();
+            })
+        );
+
+    return teams
+        .doOnNext(t -> {
+          if (t.getRight() instanceof Map) {
+            Map<String, String> response = (Map<String, String>) t.getRight();
```

Comment:
We'd like to see a cleaner approach here, let's say this:
1) we filter the roles we have and if there are no subjects matching github teams, we can skip teams fetching altogether
2) the same applies to organizations, if none are present in roles, we don't have to query organizations either

Reply:
I am not sure, if I understand, can you explain a little bit more please? Thanks!